snews_cs/snews_coinc.py

```diff
@@ -76,140 +61,95 @@ def reset_df(self):
+            self.in_coincidence = True
             return 'ALREADY_IN_LIST',
-        elif any(0 < delta_ts <= self.coinc_threshold):
+
+        # compare the current nu time with all other on the sublist
+        message_nu_time = self.times.str_to_datetime(message['neutrino_time'], fmt='%y/%m/%d %H:%M:%S:%f')
+        nu_times = pd.to_datetime(sub_list.neutrino_time, format='%y/%m/%d %H:%M:%S:%f')
+        delta_ts = ((message_nu_time - nu_times).dt.total_seconds()).values # numpy array
+
+        # if any delta t is negative, current message has to be made initial, before!
+        p_dt = delta_ts[delta_ts >= 0]
+        n_dt = delta_ts[delta_ts < 0]
+        if any(-n_dt <= self.coinc_threshold):
+            self.in_coincidence = True
+            return 'EARLY_COINCIDENT', np.insert(np.sort(np.abs(delta_ts)), 0, 0)
+        # if not check if current message is in the list
+        elif any(p_dt <= self.coinc_threshold):
+            self.in_coincidence = True
             return 'COINCIDENT', delta_ts[0]
-        elif any(0 > delta_ts >= self.coinc_threshold):
-            return 'EARLY_COINCIDENT', list(delta_ts.abs()).insert(0, 0)
-        else:
+        elif not any(np.abs(delta_ts) <= self.coinc_threshold):
+            # this condition is a must!
+            # otherwise it duplicates because it can satisfy two conditions at the same time
             return 'NOT_COINCIDENT',
+        else:
+            raise "Something weird happenning"
 
     # ------------------------------------------------------------------------------------------------------------------
     def check_coincidence_melih_seb_edition(self, message):
 
         subs_list_nums = list(self.cache_df['sub_list_num'].unique())
-        nu_times = list(pd.to_datetime(self.cache_df.neutrino_time, format='%H:%M:%S:%f'))
+        self.in_coincidence = False
+        in_list_already = False
         for sub_list in subs_list_nums:
             coinc_with_list = self.coincident_with_whole_list(message, sub_list)
             if coinc_with_list[0] == 'ALREADY_IN_LIST':
+                in_list_already = True
                 continue
             elif coinc_with_list[0] == 'COINCIDENT':
                 delta_t = coinc_with_list[1]
                 self.append_message_to_df(message, delta_t, sub_list)
+
             elif coinc_with_list[0] == 'EARLY_COINCIDENT':
                 self.append_message_to_df(message, 0, sub_list)
-                sub_list = self.cache_df.query(f'sub_list_num=={sub_list}')
-                sub_list['nu_delta_t'] = coinc_with_list[1]
-                sub_list = sub_list.sort_value(by='neutrino_time')
-                self.cache_df = pd.concat([sub_list, self.cache_df.query(f'sub_list_num!={sub_list}')],
+                _sub_list = self.cache_df.query(f'sub_list_num=={sub_list}')
+                _sub_list = _sub_list.sort_values(by='neutrino_time') # first sort! dt's are sorted too
+                # re-assign dt making the early message first
```

Comment:
here we need to do the sorting first, so that the EARLY_COINCIDENT is the first index, and the returned `delta_times` in `coinc_with_list[1]` are also already wrt earliest and they are also sorted thus all works.